Rename even more XML attributes

nrwahl2 opened the pull request:
This includes #3305. Probably will add more while killing time before Monday.

nrwahl2 commented on include/crm/msg_xml.h:
```diff
@@ -257,7 +258,7 @@ extern "C" {
 #  define XML_ACL_TAG_GROUP		"acl_group"
 #  define XML_ACL_TAG_ROLE		"acl_role"
 #  define XML_ACL_TAG_PERMISSION	"acl_permission"
-#  define XML_ACL_TAG_ROLE_REF 		"role"
```
TODO: Fix this. This isn't an attribute and it should get a PCMK_XE_ROLE.

nrwahl2 commented on daemons/based/based_notify.c:
```diff
@@ -127,7 +127,7 @@ cib_notify_send(const xmlNode *xml)
 attach_cib_generation(xmlNode *msg, const char *field, xmlNode *a_cib)
 {
-    xmlNode *generation = create_xml_node(NULL, XML_CIB_TAG_GENERATION_TUPPLE);
+    xmlNode *generation = create_xml_node(NULL, PCMK__XE_GENERATION_TUPLE);
```
Probably good for tracing only... I don't know if any callbacks use the <generation_tuple> data.

nrwahl2 commented on daemons/controld/controld_join_client.c:
```diff
@@ -146,7 +146,7 @@ void
 {
     char *join_id = user_data;
-    xmlNode *generation = create_xml_node(NULL, XML_CIB_TAG_GENERATION_TUPPLE);
+    xmlNode *generation = create_xml_node(NULL, PCMK__XE_GENERATION_TUPLE);
```
It doesn't look like the DC does anything with `<generation_tuple>` either.

nrwahl2 commented on include/crm/msg_xml_compat.h:
```diff
@@ -880,6 +880,9 @@ extern "C" {
 #define XML_ACL_TAG_ROLE_REFv1 "role_ref"
 
+//! \deprecated Do not use
+#define XML_ACL_TAG_READ "read"
```
This does have a replacement, but it's not exact -- the old name incorrectly referred to a tag, while the tag was deprecated in 1.1.12. The new symbol correctly refers to a value.

nrwahl2 commented on lib/cib/cib_ops.c:
```diff
@@ -731,7 +731,7 @@ cib_process_create(const char *op, int options, const char *section, xmlNode * r
     }
 
-    failed = create_xml_node(NULL, XML_TAG_FAILED);
+    failed = create_xml_node(NULL, PCMK__XE_FAILED);
```
another "throwaway" constant -- just one failure info XML node that needs a name

nrwahl2 commented on lib/common/ipc_controld.c:
```diff
@@ -632,7 +632,7 @@ create_hello_message(const char *uuid, const char *client_name,
     }
 
-    hello_node = create_xml_node(NULL, XML_TAG_OPTIONS);
+    hello_node = create_xml_node(NULL, PCMK__XE_OPTIONS);
```
ditto (dummy name, never checked or exposed)

nrwahl2 commented on lib/cib/cib_ops.c:
```diff
@@ -684,7 +684,7 @@ update_results(xmlNode *failed, xmlNode *target, const char *operation,
 
         was_error = true;
-        xml_node = create_xml_node(failed, XML_FAIL_TAG_CIB);
+        xml_node = create_xml_node(failed, PCMK__XE_FAILED_UPDATE);
```
ditto

nrwahl2 commented on include/crm/common/options.h:
```diff
@@ -129,6 +129,7 @@ extern "C" {
  * Enumerated values
  */
 
+#define PCMK_VALUE_CIB_BOOTSTRAP_OPTIONS    "cib-bootstrap-options"
```
As an XML ID, "VALUE" seems like the most suitable place

nrwahl2 commented on daemons/pacemakerd/pcmkd_messages.c:
```diff
@@ -110,7 +110,7 @@ handle_shutdown_request(pcmk__request_t *request)
                        "ack", NULL, CRM_EX_INDETERMINATE);
 
-    shutdown = create_xml_node(NULL, XML_CIB_ATTR_SHUTDOWN);
+    shutdown = create_xml_node(NULL, PCMK__XE_SHUTDOWN);
```
ditto (dummy name)

nrwahl2 commented:
This is so huge that GitHub refuses to show the first commits, so it would likely be impossible to review. Branching off.

nrwahl2 commented on doc/sphinx/Pacemaker_Explained/cluster-options.rst:
```diff
@@ -184,6 +184,16 @@ holds. So the decision was made to place them in an easy-to-find location.
+          pair: execution-date; cib
+       
+       execution-date
+     - :ref:`nonnegative integer <nonnegative_integer>`
```
64-bit, not 32-bit, but best I can do without overhauling the option types.

kgaillot replied:
We need a "seconds since epoch" type (some other tables that haven't been converted to specify types yet will need it too)

nrwahl2 commented on lib/pengine/pe_output.c:
```diff
@@ -1901,7 +1901,7 @@ node_xml(pcmk__output_t *out, va_list args) {
                 break;
             case node_ping:
-                node_type = "ping";
+                node_type = PCMK__VALUE_PING;
```
We could make this public since it may appear in API output and is in the API schema as an enumerated value. I'm on the fence. It **shouldn't** appear, given that it's deprecated. If it does appear, I don't know that anyone would need to fetch "ping" nodes specifically. Seems like they could just do "if not member and not remote" since there's also an "unknown" value.

nrwahl2 commented on lib/pengine/tests/rules/pe_cron_range_satisfied_test.c:
```diff
@@ -51,98 +51,130 @@ time_satisfies_year_spec(void **state) {
 time_after_year_spec(void **state) {
     run_one_test("2020-01-01",
-                 "<date_spec " PCMK_XA_ID "='spec' years='2019'/>",
+                 "<" PCMK_XE_DATE_SPEC " "
```
Ken might not like this, I dunno. Depends on whether we consider want these tests to implicitly test the values of the string constants, or make sure the functions work correctly given the values of the string constants.

nrwahl2 commented on lib/pengine/pe_output.c:
```diff
@@ -1898,6 +1898,32 @@ health2text(int health)
+ * \retval \c "unknown" otherwise
+ */
+static const char *
+node_type2str(enum node_type type)
```
Same commit as PCMK_XA_MAINTENANCE, just so that we can move node_type to the bottom of the variables when we add the final string constant for `node_xml()`.

nrwahl2 commented on tools/crm_ticket.c:
```diff
@@ -305,7 +305,7 @@ print_ticket(pcmk_ticket_t *ticket, bool raw, bool details)
             }
             fprintf(stdout, "%s=", name);
-            if (pcmk__str_any_of(name, "last-granted", "expires", NULL)) {
+            if (pcmk__str_any_of(name, PCMK_XA_LAST_GRANTED, "expires", NULL)) {
```
I don't see "expires" set anywhere, nor documented, nor in the schemas. I made a note to self to try to track down its origin.

nrwahl2 commented on tools/crm_resource_print.c:
```diff
@@ -776,7 +776,11 @@ static void
 {
     if (rsc->children == NULL) {
-        out->list_item(out, "resource", "%s", rsc->id);
+        /* Sometimes PCMK_XE_RESOURCE might act as a PCMK_XA_NAME instead of an
```
good to go back and check whether, perhaps, `--xml-simple-list` is forced for all command lines that lead to this point

nrwahl2 commented on lib/pengine/native.c:
```diff
@@ -976,15 +976,13 @@ pe__resource_xml(pcmk__output_t *out, va_list args)
     GList *only_node G_GNUC_UNUSED = va_arg(args, GList *);
     GList *only_rsc = va_arg(args, GList *);
 
+    int rc = pcmk_rc_no_output;
```
This commit does some rearranging to get variables in the same order that they're set as XML attributes.

nrwahl2 commented:
@kgaillot Based on the current naming scheme, most or all of the `PCMK_ACTION_*` constants should be `PCMK_VALUE_*`, since they're often used as the value of an XML attribute. I'm fine with keeping them as `PCMK_ACTION_*` both due to precedent and for specificity, though specificity is a slippery slope. (Thought about this while looking at a new `PCMK_ACTION_METADATA` for stonith actions -- not to be confused with `PCMK_ACTION_META_DATA` for OCF agents.)

Similarly, in `pe_output.c:cluster_options_xml()`, several of the `PCMK_OPT_*` constants (usually treated as names in an nvpair) are used as XML attribute names. I've got tentative commits that add `PCMK_XA_* equivalents to use in that context for consistency. However, I'm on the fence about it.

nrwahl2 commented:
This is the rest of them. We can either split them up again or do a marathon PR to finish them. This is mostly new constants for magic strings.

I've got some existing comments, some of which are outdated and others are still relevant. I'd especially point to https://github.com/ClusterLabs/pacemaker/pull/3308#issuecomment-1891069059.

nrwahl2 commented on lib/pengine/pe_output.c:
```diff
@@ -1143,9 +1145,7 @@ cluster_options_xml(pcmk__output_t *out, va_list args) {
-                                 PCMK_OPT_STONITH_ENABLED,
-                                 pcmk__flag_text(scheduler->flags,
-                                                 pcmk_sched_fencing_enabled),
+                                 PCMK_XA_STONITH_ENABLED, stonith_enabled,
```
I'm on the fence about this series of changes. All the XML attributes in this function call are the names of cluster options. Previously they began with `PCMK_OPT`. This series of commits adds `PCMK_XA` equivalents to use in the XML attribute context. (`PCMK_OPT` is used as the name of an nvpair.)

I could be convinced to leave these as just `PCMK_OPT`, even here.

nrwahl2 commented on include/crm/msg_xml_compat.h:
```diff
@@ -11,6 +11,7 @@
 #  define PCMK__CRM_MSG_XML_COMPAT__H
 
 #include <crm/common/agents.h>      // PCMK_STONITH_PROVIDES
+#include <crm/common/xml_names.h>
```
This include is now explicit

kgaillot commented:
Let's separate out ~50 commits at a time

kgaillot commented:
Just in case you were dreading being finished with this project: some pre-existing `XA`/`XE` constants do not match their literals (e.g. `PCMK__XA_ATTR_FORCE`), and some  of the `PCMK__VALUE_*` constants should be public

nrwahl2 commented:
> Just in case you were dreading being finished with this project: some pre-existing `XA`/`XE` constants do not match their literals (e.g. `PCMK__XA_ATTR_FORCE`), and some of the `PCMK__VALUE_*` constants should be public

Haha yep I've got notes for those.

There will still be even more magic string constants to add after these. This finishes the ones in public API output, after I rebase and whatnot.

Plus the constants in other libraries, including maybe things like `CRM_OP_*` and `F_*`.

nrwahl2 commented on include/crm/msg_xml.h:
```diff
@@ -10,446 +10,8 @@
-#  include <crm/common/xml.h>
-
 #if !defined(PCMK_ALLOW_DEPRECATED) || (PCMK_ALLOW_DEPRECATED == 1)
 #include <crm/msg_xml_compat.h>
```
@kgaillot Highlighting this, since there's so much clutter in this PR now.

nrwahl2 commented on include/crm/common/xml_names.h:
```diff
@@ -0,0 +1,455 @@
+#ifndef PCMK__CRM_COMMON_XML_NAMES__H
+#  define PCMK__CRM_COMMON_XML_NAMES__H
+
+#include <crm/common/xml.h>
```
I think we should drop this if we're allowed to. The potential problem is that it would break code any external callers that are depending on the transitive include.

I don't see anything in this file that explicitly depends on `xml.h`. However, the `ID()` macro (which may not properly belong here) depends on `nvpair.h`.

nrwahl2 replied:
In fact, if we were to drop this include, then we could include `xml_names.h` in `xml.h` and avoid the literal uses of "master" and "clone" in `crm_map_element_name()`.

nrwahl2 replied:
Should the `xml.h` include within `msg_xml.h` go inside the `PCMK__ALLOW_DEPRECATED` guard or stay outside it?

kgaillot replied:
We can actually move the include to msg_xml_compat.h. We could even prevent internal files from including msg_xml.h by adding something like this in the compat block:

    #else
    #error Include xml.h instead of msg_xml.h

By that point we don't even really need a separate compat header, but it makes sense to keep it for consistency.

nrwahl2 replied:
The include is already in `msg_xml_compat.h`, so this would just mean dropping it from `msg_xml.h` and adding the `#error` directive. Works for me.

nrwahl2 replied:
Works for me

kgaillot commented on include/crm/common/options_internal.h:
```diff
@@ -189,7 +189,6 @@ bool pcmk__valid_stonith_watchdog_timeout(const char *value);
 #define PCMK__VALUE_CUSTOM                  "custom"
 #define PCMK__VALUE_EN                      "en"
 #define PCMK__VALUE_EPOCH                   "epoch"
-#define PCMK__VALUE_GREEN                   "green"
```
commit message: presumably 7be49a9baa2?

nrwahl2 replied:
Thanks, weird.

kgaillot reviewed (commented):
Nice

nrwahl2 commented on include/crm/common/xml_names.h:
```diff
@@ -0,0 +1,454 @@
+#endif
+
+/* This file defines constants for various XML syntax (mainly element and
+ * attribute names).
```
I believe this was copy-pasted but I agree, I can reword this

nrwahl2 commented on lib/pacemaker/pcmk_simulate.c:
```diff
@@ -389,7 +389,7 @@ pcmk__profile_dir(const char *dir, long long repeat,
         char buffer[FILENAME_MAX];
 
-        out->begin_list(out, NULL, NULL, "Timings");
+        out->begin_list(out, NULL, NULL, "timings");
```
`substitutions` is used only if `--xml-substitute` is passed on the command line, which happens by default only for `crm_simulate`, `crmadmin`, `crm_mon`, and `crm_resource`.

That's fine for the `crm_simulate` schema mismatch, since `crm_simulate` is covered... I don't love the idea of outputting the capitalized version for general API output though.

kgaillot replied:
Actually the API output and tool XML output do share the same schema (all starting at api-result.rng). I don't remember why `--xml-substitute` is conditional (@clumens?); the output should always conform to the schema.

clumens replied:
No, I don't remember this anymore and the commit message isn't helpful either.  I'm sure it made sense at the time.  I wonder if I was thinking that all tools should use `--xml-substitute` when converted to using formatted output, and having it conditional was to deal with tools that had been converted to use only the new argument processing but not yet formatted output.  If that's the case, somewhere along the way I/we forgot to add it when we converted tools.

kgaillot replied:
We did make it a hidden option, so that it would be considered internal. Let's kill it and always follow the behavior, if it doesn't break existing schemas or regression tests.

nrwahl2 replied:
> all starting at api-result.rng)

Makes sense. I guess I was thinking "nothing specifically ties the `pcmk_profile_dir()` API output to the `crm_simulate*.rng` subset of the `api-result*.rng` schema." Which is still true...

With that being said, to your point, the API output wouldn't validate against the api-result schema in its current state.

> Let's kill it and always follow the behavior, if it doesn't break existing schemas or regression tests.

Sounds good to me.

nrwahl2 replied:
Seems like we don't even need the substitutions array at that point, and instead should replace them all at the source.

nrwahl2 replied:
> If there's other things that need to be validated, it would be helpful to add that somewhere.

I don't quite understand what you mean here. I know that the commands' XML output gets validated against the schema. My points were:
* the C API output isn't tied to a particular sub-schema, only to `api-result*.rng` (but we can't test it directly through a bash script)
* currently it wouldn't even validate against `api-result*.rng` for this particular function

clumens replied:
What I mean is:  we should have some way of running the C API so that its output can be validated as well.  At the moment, I don't think we're really doing anything like this, unless incidentally one of the API functions is used in a cli tool that's getting run through cts-cli.

nrwahl2 replied:
That sounds reasonable, if perhaps painful and often redundant with the CLI output. Ideally the C API should be tested independently, I agree. It's tested only indirectly via the CLI, and it's not one-to-one.

nrwahl2 commented on daemons/attrd/attrd_attributes.c:
```diff
@@ -14,11 +14,11 @@
 #include <crm/common/results.h>
 #include <crm/common/strings_internal.h>
 #include <crm/common/xml.h>
+#include <crm/common/xml_names.h>
```
I thought you didn't like transitive includes. Maybe since these are so closely related. Works for me.

nrwahl2 replied:
I'm going to get confirmation on that before I go through and change these... we can even take the stance "never include `xml_names.h` or `xml_names_internal.h` directly" if we want.

A lot of these files previously included only `msg_xml.h` and not `xml.h`. I get the feeling they were relying on THAT transitive include -- `msg_xml.h` and `xml.h` previously included each other circularly.

kgaillot replied:
Transitive includes are a gray area. It's more of a concern in headers than in source files, to guarantee that the header is always usable without the source files having to know to include random other stuff they might not directly need. Source files aren't as big a deal since only the one file can be affected. I'd still prefer to avoid unrelated transitive includes (e.g. not relying on some random header to include stdbool.h), but even then it's not horrible. For related includes between internal headers, it's not important.

Eventually I'd like to get Pacemaker to where we have a single public header and single internal header per library (much like glib has glib.h to include everything). Then we can more consistently avoid other transitive includes without too much mess.

kgaillot replied:
> I'm going to get confirmation on that before I go through and change these... we can even take the stance "never include `xml_names.h` or `xml_names_internal.h` directly; instead always include `xml.h`" if we want.

For convenience, I have followed that approach when separating one header from another, since so many files already include the original header. It's just a quick hack along the way to the more defined header-per-library approach.

> A lot of these files previously included only `msg_xml.h` and not `xml.h`. I get the feeling they were relying on THAT transitive include -- `msg_xml.h` and `xml.h` previously included each other circularly.

Yeah I fought circular includes for a long time. xml.h never included msg_xml.h though AFAIK.

clumens replied:
Personally, I like it when a header file includes everything it needs for itself - if the header uses uint32_t, it should include stdint.h.  If it uses GList, it should include glib.h.  That way it's not on the user of the header file to figure it out.  I am also fine with the one giant header that includes all the others so I have to remember less.

nrwahl2 replied:
> xml.h never included msg_xml.h though AFAIK.

In the main branch, they each include each other.

kgaillot replied:
Ah, I missed that when it was recently added (94a1afce)

nrwahl2 commented:
Addressed some review in this push. Leaving as draft to avoid unnecessary builds, as there are still a couple of pending questions.

About to rebase on current main.

kgaillot commented on include/crm/common/xml_internal.h:
```diff
@@ -240,6 +240,20 @@ pcmk__xml_artefact_root(enum pcmk__xml_artefact_ns ns);
+ * \internal
+ * \brief Retrieve the value of the \c PCMK_XA_ID XML attribute
+ *
+ * \param[in] xml  XML node to check
```
s/node/element/ since this is not intended for other kinds of nodes

kgaillot commented on lib/common/patchset.c:
```diff
@@ -507,7 +510,9 @@ process_v1_additions(xmlNode *parent, xmlNode *target, xmlNode *patch)
     CRM_CHECK(pcmk__xe_is(target, name), return);
-    CRM_CHECK(pcmk__str_eq(ID(target), ID(patch), pcmk__str_casei), return);
+    CRM_CHECK(pcmk__str_eq(pcmk__xe_id(target), pcmk__xe_id(patch),
+                           pcmk__str_casei),
```
can use none

nrwahl2 replied:
It's v1 so I was leaving it alone, but might as well

kgaillot commented:
> How do you feel about the `PCMK__{ATTRD,CONTROLD}_CMD` constants in `crm_internal.h`? The IPC code uses all these as XML attribute values, so they could reasonably become `PCMK__VALUE`.
> 
> Or they can remain separate. Because by the above logic, the `PCMK_ACTION_*` constants could also become `PCMK_VALUE_*`... depends how far we want to take this.

I think the CMDs are logically an independent class of entities rather than option/XML values. They're commands/ops/whatever that the daemons accept, via IPC and/or CPG, which happen to be implemented with XML, but could conceivably be implemented some other way without changing their purpose. ACTIONs are even more clear, since at least some of them can be passed to resource agents or in other contexts such as command-line arguments.

kgaillot reviewed (commented):
Almost done :)

nrwahl2 commented:
> > How do you feel about the `PCMK__{ATTRD,CONTROLD}_CMD` constants in `crm_internal.h`? The IPC code uses all these as XML attribute values, so they could reasonably become `PCMK__VALUE`.
> > Or they can remain separate. Because by the above logic, the `PCMK_ACTION_*` constants could also become `PCMK_VALUE_*`... depends how far we want to take this.
> 
> I think the CMDs are logically an independent class of entities rather than option/XML values. They're commands/ops/whatever that the daemons accept, via IPC and/or CPG, which happen to be implemented with XML, but could conceivably be implemented some other way without changing their purpose. ACTIONs are even more clear, since at least some of them can be passed to resource agents or in other contexts such as command-line arguments.

IDK that it's much more clear than anything else starting with `PCMK_[_]VALUE` that isn't a configurable option value. Without drawing a hard line at option values, it becomes pretty fuzzy what is a logically an independent class of entities.

With that said, most of these groups I have no strong feelings on, so I'll just bring them up to you and let you decide. I'm making notes in the background but there are more XEs and XAs to deal with first.

kgaillot commented:
> > > How do you feel about the `PCMK__{ATTRD,CONTROLD}_CMD` constants in `crm_internal.h`? The IPC code uses all these as XML attribute values, so they could reasonably become `PCMK__VALUE`.
> > > Or they can remain separate. Because by the above logic, the `PCMK_ACTION_*` constants could also become `PCMK_VALUE_*`... depends how far we want to take this.
> > 
> > 
> > I think the CMDs are logically an independent class of entities rather than option/XML values. They're commands/ops/whatever that the daemons accept, via IPC and/or CPG, which happen to be implemented with XML, but could conceivably be implemented some other way without changing their purpose. ACTIONs are even more clear, since at least some of them can be passed to resource agents or in other contexts such as command-line arguments.
> 
> IDK that it's much more clear than anything else starting with `PCMK_[_]VALUE` that isn't a configurable option value. Without drawing a hard line at option values, it becomes pretty fuzzy what is logically an independent class of entities.
> 
> With that said, most of these groups I have no strong feelings on, so I'll just bring them up to you and let you decide. I'm making notes in the background but there are more XEs and XAs to deal with first.

It's a judgment call. It would be easy to say that any magic string should be a VALUE, but that doesn't feel right. I lean to VALUE for something specifiable by users, and when its purpose is to be the value of something (vs. a more abstract concept that is used in other contexts and just happens to be passed as a value). I lean to a separate prefix when it would be helpful to see the "type" when reading the code.

Thinking about the future, the `CMD` constants in particular should be unified rather than daemon-specific. There will probably eventually be a few that all daemons support via common code. I'm thinking we should standardize on "request" terminology for these (as opposed to "op", "task", or "command", which have all been used interchangeably). The daemon can reject it, so request is better than command, and we use op and task a lot for different purposes elsewhere, but request is pretty much only used for this purpose. We can follow the literal (like `PCMK__REQUEST_PEER_REMOVE`).

nrwahl2 commented:
> It's a judgment call. ...
> ...
> ...We can follow the literal (like `PCMK__REQUEST_PEER_REMOVE`).

That sounds good to me.

I've got nothing else to add to this PR, so review at your convenience and then I'll push the next batch when this is merged.
